packages/@glimmer/application/src/application.ts

```diff
@@ -208,18 +256,25 @@ export default class Application implements Owner {
     // Create an empty root scope.
     let dynamicScope = new DynamicScope();
 
-    // The cursor tells the template which element to render into.
-    let cursor = {
-      element: (this.document as Document).body,
-      nextSibling: null
-    };
+    let builder = this.builder.getBuilder(env);
+    let templateIterator = this.loader.getTemplateIterator(this, env, builder, dynamicScope, self);
 
-    return mainLayout.renderLayout({
-      env,
-      self,
-      dynamicScope,
-      builder: clientBuilder(env, cursor)
-    });
+    // Begin a new transaction. The transaction stores things like component
+    // lifecycle events so they can be flushed once rendering has completed.
+    env.begin();
+
+    let result = this.renderer.render(templateIterator);
+
+    if (result && result.then) {
+      result.then(() => {
+        env.commit();
+        this._didRender();
+      });
+    } else {
```

Comment:
I think we might need to guard all of this with `if (result) { ... }`

Reply:
The guard is in `SyncRenderer`; the `Application` no longer knows about the render result.